[WIP] Migrate message['subject'] to message['topic']

mounilKshah opened the pull request:
**What does this PR do?**
This commit migrates the ZT to newer API elements. Here, the migration is from message['subject'] to message['topic'].
Partial fix for #965

Discussion:
[#zulip-terminal>Update to modern API elements (support v2.1) #T965](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Update.20to.20modern.20API.20elements.20.28support.20v2.2E1.29.20.23T965)

**Tested?**
- [x] Manually
- [x] Existing tests (adapted, if necessary)
- [ ] New tests added (for any new behavior)
- [ ] Passed linting & tests (each commit)

**Commit flow**
- First commit updates message element from `subject` to `topic`
- Second commit refactors the tests to accommodate this migration

**Notes & Questions**
- Fixes item 4 on #965

neiljp commented on zulipterminal/api_types.py:
```diff
@@ -21,8 +21,8 @@ class PrivateComposition(TypedDict):
     content: str
-    to: str  # stream name  # TODO: Migrate to using int (stream id)
-    subject: str  # TODO: Migrate to using topic
+    to: Optional[int]
```
This is a separate migration. It may be OK in this PR, but certainly in another commit.

neiljp commented on zulipterminal/api_types.py:
```diff
@@ -21,8 +21,8 @@ class PrivateComposition(TypedDict):
-    to: str  # stream name  # TODO: Migrate to using int (stream id)
-    subject: str  # TODO: Migrate to using topic
+    to: Optional[int]
+    topic: str
```
As per a comment in the stream, it may be cleaner to separate the compose and message parts of the topic migration into separate commits (with associated test changes in each).

neiljp commented on zulipterminal/api_types.py:
```diff
@@ -34,7 +34,9 @@ class Message(TypedDict, total=False):
     timestamp: int
     client: str
-    subject: str  # Only for stream msgs.
+    subject: str
+    # Changing subject -> topic as per the migration in Zulip 2.0
+    topic: str  # Only for stream msgs.
```
I'd suggest we comment out the subject, with a trailing comment referring to the migration.
In an ideal world that would mean using subject would cause a type error, but currently this TypedDict has total=False, so that won't happen (yet).

neiljp commented on zulipterminal/api_types.py:
```diff
@@ -53,7 +55,7 @@ class Message(TypedDict, total=False):
     content_type: str
     match_content: str  # If keyword search specified in narrow params.
-    match_subject: str  # If keyword search specified in narrow params.
+    match_topic: str  # If keyword search specified in narrow params.
```
Do you have support for this migration?

mounilKshah replied:
No. Actually, as `subject` was changed to `topic`, I felt it would be good to rename this accordingly. But as the API is still the same, I'll retain the original names for this and `orig_topic`

neiljp commented on zulipterminal/api_types.py:
```diff
@@ -145,8 +147,8 @@ class UpdateMessageEvent(TypedDict):
     message_ids: List[int]
-    orig_subject: str
-    subject: str
+    orig_topic: str
+    topic: str
```
Do you have support for this migration in the API?

neiljp reviewed (commented):
@mounilKshah You have the right ideas here, but as per my stream feedback, it's good to keep code changes and associated test changes in the same commit.

In addition, you've got multiple changes here, which ideally would be in different commits. Part of that is that the main subject->topic migration is split into Message and Composition parts, ie. what we receive from the server vs what a user sends. However, the other aspect are API changes that have similar names but haven't necessarily changed, or that are adjacent. It's fine to make these changes, but it would be cleaner to see each change in a separate commit.

Each API change may be in a different version, so that opens up the scope for writing differently about the change, when it occurred, etc. in the commit text body.

mounilKshah committed:
refactor: Migrate subject -> topic.

This commit contains migration for message["subject"]
to message ["topic"] as per the modern API with the oldest
support being v2.1. It also contains the updated tests for
the migration.

neiljp commented:
@mounilKshah I left some answers in the stream; do you need further feedback before the tests are updated?

neiljp commented:
@mounilKshah Thanks for investigating this!

I left a message in the stream, but for other readers also:
- The API relating to migration of the subject field to topic is the only (or almost only) renaming of this kind which it seems has *not* yet occurred.
- Since the only current commit in this PR relates to this, I propose we hold off on this migration, since this changes our internal data structures unnecessarily.
- Another part related to this migration is when sending messages via Composition objects, which supported topic since 2.0.0; let's redirect that effort into another PR.
